File: 1001-1500/1470.Shuffle the Array.rb
# O(1) space, O(n) time

# @param {Integer[]} nums
# @param {Integer} n
# @return {Integer[]}
def shuffle(nums, n)
    for i in (0...nums.size)
        j = i
        while nums[i] >= 0
            j = j < n ? j * 2 : (j - n) * 2 + 1
            nums[i], nums[j] = nums[j], -nums[i]
        end
    end
    for i in (0...nums.size)
        nums[i] = -nums[i]
    end
    nums
end

File: 1001-1500/1470.Shuffle the Array.js
/**
 * @param {number[]} nums
 * @param {number} n
 * @return {number[]}
 */
var shuffle = function(nums, n) {
    for (var i = 0; i < nums.length; i++) {
        var j = i;
        var buffer = nums[i];
        while (buffer >= 0) {
            j = j < n ? j * 2 : (j - n) * 2 + 1;
            var tmp = buffer; buffer = nums[j]; nums[j] = -tmp;
        }
    }
    for (var i = 0; i < nums.length; i++) {
        nums[i] *= -1;
    }
    return nums;
};

